spec/mempool/gossip/p2p.qnt

```diff
@@ -0,0 +1,141 @@
+module p2p {
+    import spells.* from "./spells"
+
+    // A node identifier.
+    type NodeID = str
+
+    //--------------------------------------------------------------------------
+    // Parameters
+    //--------------------------------------------------------------------------
+
+    // The set of all possible node IDs, even those that are not initially
+    // connected to the network.
+    const NodeIDs: Set[NodeID]
+
+    // Initial network topology. A topology is defined by the set of peers each
+    // node has.
+    const InitialPeers: NodeID -> Set[NodeID]
+
+    //--------------------------------------------------------------------------
+    // State
+    //--------------------------------------------------------------------------
+    // To model network communication, each node has a queue (a list) of
+    // incoming messages. Node A sends a message to a node B by appending the
+    // message to B's queue. We use queues to model that messages arrive in
+    // order. Messages have a sender (a node ID).
+    //
+    // In the actual implementation, transaction messages are transamitted on
+    // the `Mempool` data channel of the P2P layer. Control messages are usually
+    // transmitted on other channels with different priorities. Here we model a
+    // single, reliable channel.
+    //
+    // The type variable `msg` can be instantiated on the message types of
+    // different protocols.
+    var msgs: NodeID -> List[(NodeID, msg)]
+
+    // The dynamic network topology. Each node has a set of peers that is
+    // updated when nodes join or leave the network.
+    var peers: NodeID -> Set[NodeID]
+
+    // Auxiliary definitions
+    def Msgs(node) = msgs.get(node)
+    def Peers(node) = peers.get(node)
+
+    //--------------------------------------------------------------------------
+    // A node is in the network if it has peers.
+    val nodesInNetwork = NodeIDs.filter(node => node.Peers().nonEmpty())
+    val nodesNotInNetwork = NodeIDs.exclude(nodesInNetwork)
+
+    // A node sends a message to the given set of target nodes. It updates the
+    // list of incoming messages `_msgs`.
+    pure def send(node, _msgs, targetNodes, msg) =
+        if (targetNodes.nonEmpty())
+            _msgs.updateMultiple(targetNodes, ms => ms.append((node, msg)))
+        else _msgs
+
+    //--------------------------------------------------------------------------
+    // Actions
+    //--------------------------------------------------------------------------
+
+    // A node receives one of the incoming messages from a peer and handles it
+    // according to its type.
+    action receiveFromPeer(node, handleMessage) = all {
+        require(length(node.Msgs()) > 0),
+        val someMsg = node.Msgs().head()
+        val sender = someMsg._1
+        val msg = someMsg._2
+        // We model receiving of a message as taking the head of the list of
+        // incoming messages.
+        val updatedMsgs = msgs.update(node, tail)
```

Comment:
Is `tail` a Quint function?

Reply:
Yes, it's one of the predefined functions of lists.